server/src/main/java/com/tencent/rss/server/buffer/ShuffleBuffer.java

```diff
@@ -0,0 +1,374 @@
+        }
+        // update offset
+        currentOffset = callable.call();
+        if (currentOffset >= readBufferSize) {
+          break;
+        }
+      }
+      return new ListReadInfo(currentOffset, false);
+    } else {
+      // find lastBlockId, then read from next block
+      boolean foundBlockId = false;
+      for (ShufflePartitionedBlock block : cachedBlocks) {
+        if (!foundBlockId) {
+          if (block.getBlockId() == lastBlockId) {
+            foundBlockId = true;
+          }
+          continue;
+        }
+        if (parameter instanceof byte[]) {
+          callable = new CopyShuffleDataCall((byte[]) parameter, block, currentOffset);
+        } else if (parameter instanceof List) {
+          callable = new UpdateSegmentCall((List<BufferSegment>) parameter, block, currentOffset);
+        }
+        currentOffset = callable.call();
+        if (currentOffset >= readBufferSize) {
+          break;
+        }
+      }
+      return new ListReadInfo(currentOffset, foundBlockId);
+    }
+  }
+
+  class CopyShuffleDataCall implements Callable<Integer> {
+    final ShufflePartitionedBlock block;
+    final byte[] data;
+    final int offset;
+
+    CopyShuffleDataCall(byte[] data, ShufflePartitionedBlock block, int offset) {
+      this.block = block;
+      this.data = data;
+      this.offset = offset;
+    }
+
+    @Override
+    public Integer call() throws Exception {
+      try {
+        System.arraycopy(block.getData(), 0, data, offset, block.getLength());
+      } catch (Exception e) {
+        LOG.error("Unexpect exception for System.arraycopy, length["
+            + block.getLength() + "], offset["
+            + offset + "], dataLength[" + data.length + "]", e);
+        throw e;
+      }
+      return offset + block.getLength();
+    }
+  }
+
+  class UpdateSegmentCall implements Callable<Integer> {
+    final List<BufferSegment> bufferSegments;
+    final ShufflePartitionedBlock block;
+    final int offset;
+
+    UpdateSegmentCall(List<BufferSegment> bufferSegments,
+        ShufflePartitionedBlock block, int offset) {
+      this.block = block;
+      this.offset = offset;
+      this.bufferSegments = bufferSegments;
+    }
+
+    @Override
+    public Integer call() throws Exception {
+      bufferSegments.add(new BufferSegment(block.getBlockId(), offset, block.getLength(),
+          block.getUncompressLength(), block.getCrc(), block.getTaskAttemptId()));
+      return offset + block.getLength();
+    }
+  }
+
+  // the class has info for shuffle data reading
+  class ActualReadInfo {
+    int actualReadSize;
+    long fromEventId;
+    long lastBlockId;
+
+    ActualReadInfo(int actualReadSize, long fromEventId, long lastBlockId) {
+      this.actualReadSize = actualReadSize;
+      this.fromEventId = fromEventId;
+      this.lastBlockId = lastBlockId;
+    }
+  }
+
+  class ListReadInfo {
+    int offset;
```

Comment:
What's meaning of offset?

Reply:
The offset is used to update the offset of segment